src/aiodynamo/credentials.py

```diff
@@ -0,0 +1,286 @@
+            finally:
+                self._refresher = None
+        elif refresh is Refresh.soon:
+            if self._refresher is None:
+                self._refresher = asyncio.create_task(self._refresh(http))
+        return self._metadata and self._metadata.key
+
+    def _check_refresh(self) -> Refresh:
+        if self._metadata is None:
+            return Refresh.required
+        return self._metadata.check_refresh()
+
+    async def _refresh(self, http: HTTP):
+        self._metadata = await self.fetch_metadata(http)
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def and_then(thing: Optional[T], mapper: Callable[[T], U]) -> Optional[U]:
+    if thing is not None:
+        return mapper(thing)
+    return thing
+
+
+# https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint.html
+@dataclass
+class ContainerMetadataCredentials(MetadataCredentials):
+    timeout: Timeout = 2
+    max_attempts: int = 3
+    base_url: URL = URL("http://169.254.170.2")
+    relative_uri: str = field(
+        default_factory=lambda: os.environ.get(
+            "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI", None
+        )
+    )
+    full_uri: URL = field(
+        default_factory=lambda: and_then(
+            os.environ.get("AWS_CONTAINER_CREDENTIALS_FULL_URI", None), URL
+        )
+    )
+    auth_token: Optional[str] = field(
+        default_factory=lambda: os.environ.get(
+            "AWS_CONTAINER_AUTHORIZATION_TOKEN", None
+        )
+    )
+
+    @property
+    def url(self) -> Optional[URL]:
+        if self.relative_uri is not None:
+            return self.base_url.with_path(self.relative_uri)
+        elif self.full_uri is not None:
+            return self.full_uri
+        else:
+            return None
+
+    def is_disabled(self) -> bool:
+        return self.url is None
+
+    async def fetch_metadata(self, http: HTTP) -> Metadata:
+        headers = and_then(self.auth_token, lambda token: {"Authorization": token})
+        response = await self.fetch_with_retry(
+            http=http,
+            max_attempts=self.max_attempts,
+            url=self.url,
+            headers=headers,
+            timeout=self.timeout,
+        )
+        data = json.loads(response)
+        return Metadata(
+            key=Key(
+                id=data["AccessKeyId"],
+                secret=data["SecretAccessKey"],
+                token=data.get("Token", None),
+            ),
+            expires=parse_amazon_timestamp(data["Expiration"]),
+        )
+
+
+# https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
+@dataclass
+class InstanceMetadataCredentials(MetadataCredentials):
+    timeout: Timeout = 1
+    max_attempts: int = 1
+    base_url: URL = URL("http://169.254.169.254")
+    disabled: bool = field(
+        default_factory=lambda: os.environ.get(
+            "AWS_EC2_METADATA_DISABLED", "false"
+        ).lower()
+        == "true"
+    )
+
+    def is_disabled(self) -> bool:
```

Comment:
I wonder if there's any hint to detect that code is running on an instance.

Why it matters? If someone didn't configure anything (env, cont env), then default chain metadata thing falls back to instance metadata, and it will try to access 169.254.169.254, which is a reseved link-local address...

I dunno if someone could somehow hack the client.. 
Definitely, trying to get something from this address takes a while (socket timeout or http timeout).

```
> curl -v http://169.254.169.254
*   Trying 169.254.169.254...
* TCP_NODELAY set
* Connected to 169.254.169.254 (169.254.169.254) port 80 (#0)
> GET / HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/7.64.1
> Accept: */*
>
* Empty reply from server
* Connection #0 to host 169.254.169.254 left intact
curl: (52) Empty reply from server
* Closing connection 0
```

Reply:
if you worry (and know your env), use a concrete `Credentials` (such as `ContainerMetadataCredentials`) subclass rather than `Credentials.auto()`. if you're not on AWS, you should have the env vars set so this code never even runs.